### PR TITLE
Fix for recursive copy slowness

### LINKED ISSUE #13013: copy module doesn't follow symbolic directory
##### ISSUE TYPE

Bug Report
##### COMPONENT NAME:

copy action plugin
##### ANSIBLE VERSION

2.0.1.0
##### SUMMARY

I am copying a folder, foo, which has a subfolder, foo1, which is a symbolic link to another folder say /root/myfoo1. copy module won't copy it to remote location.

### LINKED ISSUE #21513: copy module slower than expected when recursive
##### ISSUE TYPE
 - Bug Report

##### COMPONENT NAME
copy module

##### ANSIBLE VERSION
```
ansible 2.2.1.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = Default w/o overrid
```

##### OS / ENVIRONMENT
Debian Sid -> Centos 6

##### SUMMARY
The copy module documentation states:
> The “copy” module recursively copy facility does not scale to lots (>hundreds) of files.
and I'm finding incredible slowdowns at barely 50 files.

If I list each file individually using with_items in my task however, ansible completes an order of magnitude faster.

I'm confident that there is a large speed boost to be gained somewhere here.

##### STEPS TO REPRODUCE
```
copy: src=directory dest=somewhere
```
vs
```
file: path=somewhere state=directory

copy: src={{ item }} dest=directory/{{ item }}
with_items:
  - file1
  - file2
  - etc
```

##### EXPECTED RESULTS
Using the copy module recursively _should_ take a similar amount of time to listing the files using with_items?

##### ACTUAL RESULTS
Using the copy module recursively takes at least 10x longer.

### PR DESCRIPTION
**Pull request by abadger**
Copy module was walking over files in subdirectories repeatedly (a
directory tree a few levels deep could bring the time spent into the
tens of minutes)

This was traced to the fix for this bug report: https://github.com/ansible/ansible/issues/13013

Fixed #13013 a different way and added an integration test to check for
regressions of #13013 as we optimize this code.

Fixes #21513

##### ISSUE TYPE

 - Bugfix Pull Request

##### COMPONENT NAME
lib/ansible/plugins/action/copy.py

##### ANSIBLE VERSION

```
devel 2.3 2.2
```

Will only push to devel for now.  May decide to push back to 2.3 but not right before release as copy is too central to people's usage.

### CODE REVIEW

**Comment by abadger** [comment]
@PriceChild @jctanner @bcoca Would appreciate review from some or all of you on this.

**Comment by bcoca** [comment]
yeah, that looks like a total misunderstanding of the return of os.walk

**Comment by abadger** [comment]
Merged to devel and cherry-picked to temp-staging-post2.3.0 for potential inclusion in 2.3.1

**Comment by widyono-cets** [comment]
TL;DR: Forcing followlinks=True is a regression. If a tree has a circular symlink this will cause a failure of the ansible run. Furthermore, follow=no by default for copy module.

The removals in commit 78ced53 are good because they remove the redundant recursion originally introduced in trying to fix #13013. However, instead of forcing followlinks=True and ignoring the follow parameter, a proper fix for #13013 would check each dirname returned for os.path.islink(), and if follow is no, it should simply copy the symlink. If follow is yes, then do a recursion into the target of the symlink (after testing for safety with a try and Except OSError, checking for errno.ELOOP).

This way, you are honouring the follow argument to the copy module, and you are ensuring a circular symlink does not choke ansible.

**Comment by abadger** [comment]
@widyono-cets If you could open a new Issue, that would be great... I almost didn't see your comment here since we seldom look at closed and merged issues and PRs.  If you're on IRC, let me know what the new issue number is there (I'm abadger1999 in #ansible-devel on irc.freenode.net).

I'll try to work on a fix for circular symlinks and post a PR.  However, I'm not sure whether follow is supposed to be honored or not here.  I've been talking with bcoca and he's not certain either.  We'll  discuss it more but would love to have the bug opened so I have a place to record our conversations.  Thanks!

**Comment by abadger** [comment]
the conversation so far: copy seems to be special in that it has to deal with symlinks on both the controller and managed machine.  Other modules on the controller side, we always want to dereference symlinks.  On the module side, we have follow to determine whether to dereference or not.  Maybe we need a new parameter just for copy to determine whether to follow controller-side or not.

**Comment by abadger** [comment]
Big note:  Due to the circular symlink problem @widyono-cets brought up, I'm reworking the code for this for devel.  @PriceChild I also ended up not merging it to stable-2.3 after the 2.3.0 release.  It is currently not scheduled to go into 2.3.1.  It may make 2.3.2 if I can get a suitable implementation done for devel before that and then backport it to 2.3

**Comment by abadger** [comment]
Just an update here that https://github.com/ansible/ansible/issues/24393 is nearly finished.  Waiting on a code review before pushing it into devel for 2.4.0.  The code changes are very large so it's unfortunately not something I feel comfortable pushing back to 2.3.x.  So 2.4.0 will (assuming no insurmountable bugs are found) be the first release to contain the fix.
